Clean up flag values in livepeer.go

angyangie opened the pull request:
**What does this pull request do? Explain your changes. (required)**
This is the first step in this issue: https://github.com/livepeer/go-livepeer/issues/710. It is a halfway step to adding a config file and different flag parsing code to `livepeer.go`.

Generally, how we handle flags needs a lot of cleanup. Rather than flag parsing or the number of flags itself, this issue is more around how we check and use those flags in `cmd/livepeer` to initialize the node. The code interleaves between initialization for orchestrators, broadcasters and both, often switching back-and-forth.

Discussion on flag clean-up is here:
https://github.com/livepeer/go-livepeer/issues/709
https://github.com/livepeer/go-livepeer/issues/580

**Specific updates (required)**
- Grouped flags together based on general function (makes the file more readable)
- Added a `NetworkConfig` struct to hold network config options like `ethUrl`, `ethController`, and `monHost`. If more network specific options are added in the future, we could add them here.
- Removed `maxPricePerSegment` (unused flag)
- Combined `ethIpcPath` and `ethUrl`
- Renamed `noipfsLogFiles` to `logIpfs`
- Renamed `monhost` to `monHost`
- Made `transcoder` var into a flag,  removed `standaloneTranscoder`
- Removed `orchAndTrans`. Temporarily setting orchs and transc to be both, here: 
```golang
if *transcoder || *orchestrator {
		*transcoder = true
		*orchestrator = true
	}
```
- Added `-network` flag. Removed `-rinkeby`, `devenv`, `offchain` flags. If network isn't set, `-network offchain` is assumed, regardless of what other options might have been specified. If network is set, but not `rinkeby` or `mainnet`, network remains as was specified. (REVIEW)
- Edited devtool script to work with all changes


**To do:**
- [x] (TBD) Add a `-broadcaster` flag for consistency 
- [x] (Must review) No way for user to operate in `offchain` mode and to name their own network
- [x] (TBD) Allow users to set their own `datadir` name, without network subfolder?
- [ ] More testing on different networks

**Will not do:**
Remove `ipfsPath`, `noipfsLogFiles`, unused flags - a lot of code was already written in anticipation of these flags. Since it seems like these flags will be used in the near future, I vote for keeping them commented out. 

**How did you test each of these updates (required)**
I can successfully transcode using devtool. All tests passed. 

**Does this pull request close any open issues?**
https://github.com/livepeer/go-livepeer/issues/709
https://github.com/livepeer/go-livepeer/issues/580

**Checklist:**
- [X] README and other documentation updated
- [X] Node runs in OSX and devenv
- [X] All tests in `./test.sh` pass

Commit:
clean up flag values in livepeer.go

angyangie commented on cmd/livepeer/livepeer.go:
```diff
@@ -130,47 +153,30 @@ func main() {
-		if *monitor && *monhost == "" {
-			*monhost = "http://metrics-mainnet.livepeer.org/api/events"
-		}
+		*network = "offchain"
```
Note: need to add ability for users to name their own network, and still be "offchain"

j0sh replied:
Can we just make `offchain` the default in `network := flag.String("network", "", "Network to connect to")` ?

That being said, I think we can live without the "feature" of giving another name to an offchain network. The only thing that changes is the default datadir, which the user can still set to whatever.

angyangie replied:
True! We can indeed live without that feature. I'll make the default `*network = "offchain"`

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -60,67 +58,92 @@ func main() {
+		},
+	}
+
+	// temporary until we fully separate transcoder and orch nodes
```
Each mode can already be run standalone so I don't think this is needed

angyangie replied:
True. I placed it there to mirror previous functionality with `* orchAndTrans` - when transcoder was set to true, `* orchAndTrans` also became true. My concern was keeping `devtool` script working. But, to that end, I'll just add an `-orchestrator` flag to devtool script and achieve the same result.

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -130,47 +153,30 @@ func main() {
-		transcoder = true
-		*orchestrator = true
-	}
+	*datadir = *datadir + "/" + *network
```
```
import "path/filepath"
if *datadir == "" {
  *datadir = filepath.Join(usr.HomeDir, ".lpData", *network)
}
```

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -130,47 +153,30 @@ func main() {
-			if *ethUrl == "" {
-				*ethUrl = "wss://rinkeby.infura.io/ws"
+	/// Setting config options based on specified network
+	if *network != "" {
```
Don't think this `if` is necessary since configOptions[*network] will just return empty if the network is empty

angyangie replied:
That's right - now that we've set a default network of "offchain", no need to do this check.

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -130,47 +153,30 @@ func main() {
+	/// Setting config options based on specified network
+	if *network != "" {
+		if netw, ok := configOptions[*network]; ok {
+			if *orchestrator {
```
We'd want to also set these values for broadcasters. Seems OK to just set the values regardless of the node type.

angyangie replied:
Sounds good. In the original file, we first checked for `if !*offchain && !transcoder {`. Do we need transcoders to also have these values? I could edit that line to read: `if *orchestrator || *broadcaster {`, and add a broadcaster flag, which I intended to do today?

j0sh replied:
Transcoders don't really need to specify a `-network` flag. If they do, it seems harmless, so I wouldn't special case this to exclude transcoders in favor of keeping the flow straight-through.

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -60,67 +58,92 @@ func main() {
+		"rinkeby": {
+			ethUrl:        "wss://rinkeby.infura.io/ws",
+			ethController: "0x37dc71366ec655093b9930bc816e16e6b587f968",
+			monHost:       "http://metrics-rinkeby.livepeer.org/api/events",
```
Presented like this, the naming feels a bit inconsistent in having `ethUrl` and `monHost` . How about `monUrl` ?

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -254,14 +260,14 @@ func main() {
 			//Connect to specified IPC file
-			gethUrl = *ethIpcPath
+			gethUrl = *ethUrl
 		} else if *ethUrl != "" {
```
This part isn't needed anymore

angyangie replied:
Ah yes, I had that change locally. Was going to push today!

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -254,14 +260,14 @@ func main() {
 		gethUrl := ""
-		if *ethIpcPath != "" {
+		if *ethUrl != "" {
 			//Connect to specified IPC file
```
"Connect to the specified Ethereum node"

j0sh commented on cmd/livepeer/livepeer.go:
```diff
@@ -254,14 +260,14 @@ func main() {
 			gethUrl = *ethUrl
 		} else {
-			glog.Errorf("Need to specify ethIpcPath or ethWsUrl")
+			glog.Errorf("Need to specify ethUrl or ethWsUrl")
```
`glog.Error("Need to specify ethUrl")`

Commit:
edits in response to comments

j0sh committed:
livepeer: Changes around the datadir CLI flag.

* Fix datadir default value. Would not have set the network
  sub-folder otherwise.

* Use HOME environment variable for homedir where possible.
  Facilitates custom homedirs for testing.

* Create intermediate directories for the datadir. This casued
  issues with the ~/.lpData/<network-name> scheme where we would
  have to create the <network-name> directory but but the
  operation would fail if ~/.lpData did not exist first.

j0sh committed:
livepeer: Improve handling of required O/T flags.

j0sh committed:
livepeer: Make the node type mandatory.

j0sh committed:
livepeer: Consolidate B's O Discovery setup into earlier code blocks.

j0sh committed:
broadcaster: Fix result sig check in offchain mode.

j0sh committed:
test: Add basic checking of cli flags.

j0sh commented:
Pushed some changes.

* Node type (-broadcaster / -transcoder / -orchestrator) is now **required** and will fail with a message if not specified. Earlier versions of this PR did not set a type if none was omitted. Whether we should have a default (broadcaster?) is certainly up for discussion. 356f87a83b9e1334ac7d9d7f7cd4d8ae4b05da36

* Improved handling of required O/T flags. 9436176d8187f7e599a2f3aa81d0f6f470177dea

* Look at the $HOME environment variable prior to asking the OS for the user's homedir. This is better practice anyway, and facilitates testing. Since it's technically non-portable, we fall back to the old method if $HOME is not set. 6a7807615973f9621e8dbda31ad63bbe568b6255

* Some other code consolidation around B's flag handling 453741dfae1ce90264f4f28ebc27199a725fd5fb

* Fix B's result sig checking in offchain mode. Separate bug but including it here anyway. 21bcb43aba89e4a5a9fb01fdfdbb63f0c134e0e1

* Added a test script that checks our handling of CLI flags. 4a167d3f9770d628aff65c86cbcf15b72c60665f

Once this is merged, we need to make the corresponding changes to devenv and the test harness.

j0sh committed:
readme: Update.

j0sh commented:
Updated the README and rebased onto master. 94b30b595be8ada3f0e83a2d4f1863ba0b542070

Since there are some behavioral changes (2530428a312244b8d56ecb8fea65f71a22a0a129 , f34a80d23e6adf84208f152048c5d7358ac901a9) it would be nice to get some eyes on this before merging. See the [earlier comment](https://github.com/livepeer/go-livepeer/pull/738#issuecomment-466238486)  for details.

cc @angyangie @darkdarkdragon @yondonfu @ya7ya @iameli @ericxtang

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -254,14 +254,11 @@ func main() {
-			gethUrl = *ethIpcPath
-		} else if *ethUrl != "" {
-			//Connect to specified websocket
+		if *ethUrl != "" {
```
Could we simplify this if/else into:

```
if *ethUrl == "" {
    glog.Error("Need to specify ethUrl")
    return
}
```

since `*ethUrl` would either be equal to the user specified value or one of the network defaults at this point? We could also remove the extra `gethUrl` variable.

j0sh replied:
Fixed

yondonfu commented on cmd/devtool/devtool.go:
```diff
@@ -331,9 +334,9 @@ func remoteConsole(destAccountAddr string) error {
+	if !ethControllerOverride {
 		// f9a6cf519167d81bc5cb3d26c60c0c9a19704aa908c148e82a861b570f4cd2d7 - SetContractInfo event
-		getControllerAddressScript := `
+		getethControllerScript := `
```
`getethControllerScript` -> `getEthControllerScript`? Same comment applies for the other appearances of this variable in this file.

j0sh replied:
Fixed

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -397,21 +410,10 @@ func main() {
 
 	// Set up logging
-	if !*noIPFSLogFiles {
+	if !*logIPFS {
```
Perhaps it would be clearer to default `logIPFS` to true and check for `if *logIPFS {}` here?

j0sh replied:
Ah this actually should have been `if *logIPFS {`. Fixed.

Should we set the default to `true` even with IPFS being unused?

yondonfu replied:
Hm right - I think it makes sense to default to `false` right now since IPFS is unused, but if/when we start using IFPS again then it might make sense to default it to `true` to avoid IPFS log messages adding noise to the Livepeer node log output.

yondonfu commented on test_args.sh:
```diff
@@ -0,0 +1,93 @@
+[ ! -d  "$CUSTOM_DATADIR"/rinkeby ] # sanity check that network isn't included
+kill $pid
+
+# check invalid / missing service address.
```
More specifically, we'd want to test for a mismatch between the specified service address (either declared on-chain or user provided) and the inferred service address right?

j0sh replied:
It's meant to ensure an unset serviceURI makes the availability check fail, since the inferred address will (most likely!) be non-public. Updated the comment to reflect that.

One day we should spend time getting this type of test to work reliably. Then there are various items we should test related to this specific item:
* Missing `serviceAddr` (the current test, but commented out)
* Malformed `serviceAddr` (added this test)
* Well-formed `serviceAddr` pointing somewhere else
* On-chain but empty `serviceURI`
* On-chain but malformed `serviceURI`
* On-chain, well-formed `serviceURI` pointing somewhere else

yondonfu reviewed (approved):
LGTM

angyangie commented on cmd/livepeer/livepeer.go:
```diff
@@ -229,10 +224,13 @@ func main() {
 
 	if n.NodeType == core.TranscoderNode {
 		glog.Info("***Livepeer is in transcoder mode ***")
+		if n.OrchSecret == "" {
```
Good thinking here

angyangie commented on test_args.sh:
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env bash
```
Nice addition (whole file)!
